src/v/storage/readers_cache.cc

```diff
@@ -339,11 +345,33 @@ void readers_cache::dispose_in_background(entry* e) {
 
 ss::future<> readers_cache::maybe_evict() {
     auto now = ss::lowres_clock::now();
+    /**
+     * First try to evict invalid and outdated readers
+     */
     co_await evict_if([this, now](entry& e) {
         const auto invalid = !e.reader->is_reusable() || !e.valid;
         const auto outdated = e.last_used + _eviction_timeout < now;
         return invalid || outdated;
     });
+
+    // cap the leaders size
+    evict_size();
+}
+
+void readers_cache::evict_size() {
```

Comment:
Given the definition of the max size config I'm wondering if it makes sense to block new puts, or aggressively evict when putting a new reader when we surpass the max size

Reply:
I was thinking about this initially. i.e. do what usual caches done - evict on put. I was a little bit worried about the impact on put performance, also it will still have to be done in background.